### PR TITLE
Upgrade antd

### LINKED ISSUE #711: Upgrade antd to latest
## Feature Request

**Is your feature request related to a problem? Please describe:**

When upgrading antd to latest version, our CardTabs are broken.

**Describe the feature you'd like:**

Resolve the compatibility issue and upgrade antd to latest version.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### LINKED ISSUE #803: Update UI dependencies
## Feature Request

**Is your feature request related to a problem? Please describe:**

Recent Antd introduces several new handy features, like form label description (in question mark icon), success typography, etc. However now we cannot use it.

**Describe the feature you'd like:**

Update Antd (and maybe other UI libraries as well, however I suggest to open another PR for others).

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

I have tried with latest Antd shortly, and discovered that our CardTabs & CardTable have some small problems with it. So we need tiny modifications in addition to just simply upgrade the dependency.

### PR DESCRIPTION
**Pull request by baurine**
close #803 , close #711 
upgrade antd from 4.2.5 to 4.8.5

### CODE REVIEW

File: ui/lib/apps/ClusterInfo/pages/List.module.less [hunk 1]
```diff
@@ -0,0 +1,6 @@
+.sticky_tabs_header {
+  padding-left: 48px;
```
Comment (breeswish) [inline ui/lib/apps/ClusterInfo/pages/List.module.less:2]: Use variables

File: ui/lib/apps/ClusterInfo/pages/List.module.less [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+  height: 46px;
+  margin-bottom: 16px;
+  border-bottom: 1px solid #f0f0f0;
```
Comment (breeswish) [inline ui/lib/apps/ClusterInfo/pages/List.module.less:5]: Use variables

File: ui/lib/apps/ClusterInfo/pages/List.module.less [hunk 3]
```diff
@@ -0,0 +1,6 @@
+.sticky_tabs_header {
+  padding-left: 48px;
+  height: 46px;
+  margin-bottom: 16px;
```
Comment (breeswish) [inline ui/lib/apps/ClusterInfo/pages/List.module.less:4]: Use variables

File: ui/lib/apps/ClusterInfo/pages/List.module.less (hunk 1 above)
Comment (baurine) [reply ui/lib/apps/ClusterInfo/pages/List.module.less:2]: Yep, will update it, it is still wip.

**Comment by breeswish** [comment]
A kind reminder: I previously found that Tab+Table will have problems, table inside is lazily resized.. Maybe you can try it out and find out how to avoid this issue..

**Comment by baurine** [comment]
> A kind reminder: I previously found that Tab+Table will have problems, table inside is lazily resized.. Maybe you can try it out and find out how to avoid this issue..

Got, let me try it.

**Comment by baurine** [comment]
> A kind reminder: I previously found that Tab+Table will have problems, table inside is lazily resized.. Maybe you can try it out and find out how to avoid this issue..

I found this depends on the screen width, the former implementation has this problem as well. When the screen is wide, the former implementation is lazily resized at the table end while the new implementation is lazily resized at the first column of the table. (not captured as gif)

While the screen is narrow, the former implementation is lazily resized at the first column while the new implementation doesn't resize.

When the screen is narrow, the former implementation:



the new implementation:

**Comment by baurine** [comment]
Hard to dig, but I found an workaround, just use the Tab as the indicator, not the container:

```jsx
    <ScrollablePane style={{ height: '100vh' }}>
      <Card>
        <CardTabs
          defaultActiveKey={tabKey}
          onChange={(key) => {
            navigate(`/cluster_info/${key}`)
          }}
          renderTabBar={renderTabBar}
          animated={false}
        >
          <CardTabs.TabPane
            tab={t('cluster_info.list.instance_table.title')}
            key="instance"
          ></CardTabs.TabPane>
          <CardTabs.TabPane
            tab={t('cluster_info.list.host_table.title')}
            key="host"
          ></CardTabs.TabPane>
          <CardTabs.TabPane
            tab={t('cluster_info.list.store_topology.title')}
            key="store_topology"
          ></CardTabs.TabPane>
        </CardTabs>
        {tabKey === 'instance' && <InstanceTable />}
        {tabKey === 'host' && <HostTable />}
        {tabKey === 'store_topology' && <StoreLocation />}
      </Card>
    </ScrollablePane>
```

Not sure whether it is fine.

File: ui/lib/apps/ClusterInfo/pages/List.module.less [hunk 4]
```diff
@@ -0,0 +1,14 @@
+@import '~antd/es/style/themes/default.less';
+
+.sticky_tabs_header {
+  padding-left: @padding-page; // 48px
+  height: @padding-page; // 48px
+  margin-bottom: @padding-md; // 16px
+  border-bottom: 1px solid @gray-4;
```
Comment (baurine) [inline ui/lib/apps/ClusterInfo/pages/List.module.less:7]: Although `1px` and `solid` also have variables, I think here using the value may be more direct and clear.

**Comment by baurine** [comment]
Final effect:

**Comment by baurine** [comment]
hi @breeswish PTAL, thanks!

**Comment by breeswish** [comment]
Fantastic! I will try with this PR later this week.

**Comment by unbyte** [comment]
To avoid potential problems, I think it's better to check for changes in the changelog of antd. https://github.com/ant-design/ant-design/blob/master/CHANGELOG.zh-CN.md

**Comment by baurine** [comment]
> To avoid potential problems, I think it's better to check for changes in the changelog of antd. https://github.com/ant-design/ant-design/blob/master/CHANGELOG.zh-CN.md

I did, but too many 😂, and I noticed that it said the Tabs components have been refactored in 4.3.0

> 🔥 重做 Tabs 以提升多标签在不同环境下的用户体验，DOM 结构完全重写，请注意覆盖样式丢失的问题。

**Comment by breeswish** [comment]
> Hard to dig, but I found an workaround, just use the Tab as the indicator, not the container:
> 
> ```js-jsx
>     <ScrollablePane style={{ height: '100vh' }}>
>       <Card>
>         <CardTabs
>           defaultActiveKey={tabKey}
>           onChange={(key) => {
>             navigate(`/cluster_info/${key}`)
>           }}
>           renderTabBar={renderTabBar}
>           animated={false}
>         >
>           <CardTabs.TabPane
>             tab={t('cluster_info.list.instance_table.title')}
>             key="instance"
>           ></CardTabs.TabPane>
>           <CardTabs.TabPane
>             tab={t('cluster_info.list.host_table.title')}
>             key="host"
>           ></CardTabs.TabPane>
>           <CardTabs.TabPane
>             tab={t('cluster_info.list.store_topology.title')}
>             key="store_topology"
>           ></CardTabs.TabPane>
>         </CardTabs>
>         {tabKey === 'instance' && <InstanceTable />}
>         {tabKey === 'host' && <HostTable />}
>         {tabKey === 'store_topology' && <StoreLocation />}
>       </Card>
>     </ScrollablePane>
> ```
> 
> Not sure whether it is fine.
> 
> 

Nice dig, can we encapsulate this logic into our own components, in order to avoid writing this hack everywhere? Or is there "detached tabs" available in Antd?

**Comment by baurine** [comment]
> > Hard to dig, but I found an workaround, just use the Tab as the indicator, not the container:
> > ```
> >     <ScrollablePane style={{ height: '100vh' }}>
> >       <Card>
> >         <CardTabs
> >           defaultActiveKey={tabKey}
> >           onChange={(key) => {
> >             navigate(`/cluster_info/${key}`)
> >           }}
> >           renderTabBar={renderTabBar}
> >           animated={false}
> >         >
> >           <CardTabs.TabPane
> >             tab={t('cluster_info.list.instance_table.title')}
> >             key="instance"
> >           ></CardTabs.TabPane>
> >           <CardTabs.TabPane
> >             tab={t('cluster_info.list.host_table.title')}
> >             key="host"
> >           ></CardTabs.TabPane>
> >           <CardTabs.TabPane
> >             tab={t('cluster_info.list.store_topology.title')}
> >             key="store_topology"
> >           ></CardTabs.TabPane>
> >         </CardTabs>
> >         {tabKey === 'instance' && <InstanceTable />}
> >         {tabKey === 'host' && <HostTable />}
> >         {tabKey === 'store_topology' && <StoreLocation />}
> >       </Card>
> >     </ScrollablePane>
> > ```
> > 
> > 
> > Not sure whether it is fine.
> > 
> 
> Nice dig, can we encapsulate this logic into our own components, in order to avoid writing this hack everywhere? Or is there "detached tabs" available in Antd?

Let me try it.

File: ui/lib/components/CardTabs/index.module.less [hunk 5]
```diff
@@ -3,7 +3,7 @@
   margin-right: -@padding-page;
 
   :global {
-    .ant-tabs-bar {
+    .ant-tabs-nav {
```
Comment (breeswish) [inline ui/lib/components/CardTabs/index.module.less:6]: Maybe better to find some way to remove this rule (which depends on the internal implementation of Antd's Tab).

Comment (baurine) [reply ui/lib/components/CardTabs/index.module.less:6]: ok, let me try it.

File: ui/lib/components/CardTabs/index.tsx [hunk 6]
```diff
@@ -1,23 +1,51 @@
...
+}
+
 export interface ICardTabsProps extends TabsProps {
   className?: string
-  children?: React.ReactNode
+  tabs: Tab[]
 }
 
-function CardTabs({ className, children, ...restProps }: ICardTabsProps) {
+function CardTabs({
```
Comment (baurine) [inline ui/lib/components/CardTabs/index.tsx:18]: hi @breeswish , refactor the CardTabs like this, is it ok?

File: ui/lib/components/CardTabs/index.module.less (hunk 5 above)
Comment (baurine) [reply ui/lib/components/CardTabs/index.module.less]: It works!

**Comment by unbyte** [comment]
LGTM

**Review by breeswish — APPROVED** [review]

**Pull request merged**